NOAA-OWP/ras2fim
# [2pt] Fixing incorrect benchmark results and Geocurves

During testing and comparing ras2fim output units to benchmark data, we discovered some problems with geocurves. Many were not being created as expected.

Part of the evaluation of a unit is to use a new tool called `run_eval_bench_data.py`.  This is a WIP tool and requires a fair bit of hardcoding to be used at this time, but it is expected to evolve as time permits later.

The normal "alpha testing" evaluation of a unit is to run the `run_test_cases.py` against a unit, then run `run_eval_bench_data.py` to see metrics and agreement rasters.

**Note: This release does require a ras2fim Conda environment reload**

### Additions  

- `tools\run_eval_bench_data.py`: as described above.

### File Renamed
- Was: `tools\run_unit_benchmark_tests.py`,  Now: `run_test_cases.py`: This is the same name as used in the FIM product for this functionality and helps minimize confusion.

### Changes  

- `environment.yml`: Adding seaborn used for plots and updated a few other packages
- `src`
    - `conflate_hecras_to_nwm.py`: Linting fixes
    - `create_geocurves.py`: A wide number of changes to fix the bug listed above. It also has significantly upgraded logging.
    - `create_shapes_from_hecras.py`: Added a note about an import fix required later. See Issue [323](https://github.com/NOAA-OWP/ras2fim/issues/323) Change logic of filtering out models and key model files.
- `tools`
     - `ras2inundation.py`: A validation fix.
    - `run_test_cases.py`: (renamed as mentioned above): Some linting updates and some debugging cleanup
    - `acquire_and_preprocess_3dep_dems.py`: A small fix to disable the levee system in this script. The levee system is not fully operational system wide.

### Testing

A very wide array of testing was done with small set models, plus full units (all models).  

To test:
- Update your conda environment.
- Ensure you have or build a full set of models for a HUC.
- Load the latest eval_PROD_metrics.csv from S3. If you already have on downloaded, remember that the eval system will overwrite records in the metrics file that match the unit_name and unit_version.  ie) 12090301_2277_ble_240519.
- Run the `tools\run_test_cases.py`. Watch the notes and args as this is stable tool but some args have never been tested. It is encouraged to use the defaults args and pathing. This will give you metrics files and agreement files.
- Run the `tools\run_eval_bench_data.py`. It is a very new tool and is in rough prototype form. You will need to change some hardcoded values in order to use at this time, but it is very valuable to comparing results from V1 to V2 or any version of a unit.


### Checklist

_You may update this checklist before and/or after creating the PR. If you're unsure about any of them, please ask, we're here to help! These items are what we are going to look for before merging your code._

- [ ] Informative and human-readable title, using the format: `[_pt] PR: <description>`
- [ ] Pre-commit executed and linting changes made.
- [ ] Links are provided if this PR resolves an issue, or depends on another other PR
- [ ] If submitting a PR to the `dev` branch (the default branch), you have a descriptive [Feature Branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) name using the format: `dev-<description-of-change>` (e.g.: `dev-revise-levee-masking`)
- [ ] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
- [ ] Changes adhere to [PEP-8 Style Guidelines](https://peps.python.org/pep-0008/)
- [ ] Project documentation has been updated ([CHANGELOG](/doc/CHANGELOG.md) and/or [README](/README.md))
- [ ] [Reviewers requested](https://help.github.com/articles/requesting-a-pull-request-review/)

### Merge Checklist (For Technical Lead use only)

- [ ] Update [CHANGELOG](/docs/CHANGELOG.md) with latest version number and merge date
